cpg_utils/slack.py

```diff
@@ -0,0 +1,110 @@
+"""
+Sending notifications to Slack. To enable, create a channel, then add a
+corresponding Slack app (e.g. Seqr Loader) to the channel:
+
+/invite @Seqr Loader
+
+Make sure `slack/channel`, `slack/token_secret_id`, and `slack/token_project_id`
+configuration values are set.
+"""
+
+import os
+from textwrap import dedent
+import logging
+
+from hailtop.batch.job import Job
+import slack_sdk
+
+from cpg_utils.cloud import read_secret
+from cpg_utils.config import get_config
+
+
+def send_message(text: str):
+    """
+    Send text as a Slack message, reading credentials from the config.
+    """
+    if slack_channel := (
+        os.environ.get('SLACK_CHANNEL') or get_config().get('slack', {}).get('channel')
+    ):
+        slack_token = get_token()
+
+        slack_client = slack_sdk.WebClient(token=slack_token)
+        try:
+            slack_client.api_call(  # pylint: disable=duplicate-code
+                'chat.postMessage',
+                json={
+                    'channel': slack_channel,
+                    'text': text,
+                },
+            )
+        except slack_sdk.SlackApiError as err:
+            logging.error(f'Error posting to Slack: {err}')
+
+
+def get_token() -> str:
+    """
+    Returns Slack token.
+    """
+    if token := os.environ.get('SLACK_TOKEN'):
+        return token
+    token_secret_id = get_config()['slack'].get('token_secret_id')
+    token_project_id = get_config()['slack'].get('token_project_id')
+    if not token_secret_id or not token_project_id:
+        raise ValueError(
+            '`slack.token_secret_id` and `slack.token_project_id` '
+            'must be set in config to retrieve Slack token'
+        )
+    token = read_secret(
+        project_id=token_project_id,
+        secret_name=token_secret_id,
+        fail_gracefully=False,
+    )
+    assert token
+    return token
+
+
+def slack_env(j: Job):
+    """
+    Add environment variables that configure Slack reporter.
+    """
+    if not (channel := get_config().get('slack', {}).get('channel')):
+        return
+    j.env('SLACK_CHANNEL', channel)
+    j.env('SLACK_TOKEN', get_token())
```

Comment:
Do we to bind the secret in here, puts it more on the public record and binds it in logs. It's not a super serious secret token, but IMO would be better to enforce getting the token from secrets each time (costs about $3 / million requests)

Reply:
I've removed everything except for the Python-based `send_message` implementation now.